### PR TITLE
 VPR binary rr_node and rr_edge read/write

### PR DESCRIPTION
**Pull request by andrewb1999**
Fixed rebase issues

#### Description

Speed up rr_graph file reading by converting rr_nodes and rr_edges to binary.  This includes:
1) An option to write out a binary file and associated shortened .xml rr_graph that tells vpr to use the binary file for rr_nodes and rr_edges
2) The ability to read rr_graph with binary file

#### Motivation and Context

Reading SymbiFlow rr_graph took 30+ seconds each time which became a significant portion of our pack, place, and route time.  Reading xml is generally slow, so replacing the rr_nodes and rr_edges xml tags (the majority of the file) with a reference to a binary file to read drastically speeds up this reading to ~4 seconds if you do not check the rr_graph.

#### How Has This Been Tested?

We ran it on a number of our designs and had exactly the same .net .place and .route with using xml and using binary.

#### Usage

To write rr_graph .bin.xml and .bin (must have .real.xml rr_graph produced before running this): 
$VTR_DIR/vtr-verilog-to-routing/build/vpr/vpr $SYMB_DIR/symbiflow-arch-defs/build/xc7/archs/artix7/devices/xc7a50t-basys3-roi-virt/arch.timing.xml --device xc7a50t-basys3-test $SYMB_DIR/symbiflow-arch-defs/common/wire.eblif --read_rr_graph $SYMB_DIR/symbiflow-arch-defs/build/xc7/archs/artix7/devices/rr_graph_xc7a50t-basys3_test.rr_graph.real.xml --route_chan_width 6 --min_route_chan_width_hint 1 --write_rr_graph $SYMB_DIR/symbiflow-arch-defs/build/xc7/archs/artix7/devices/rr_graph_xc7a50t-basys3_test.rr_graph.real.bin.xml --pack --disable_errors check_unbuffered_edges:check_route --allow_dangling_combinational_nodes on

To read binary rr_graph:
Run VPR as you normally would but point it to your rr_graph .bin.xml file instead of the normal .xml rr_graph

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: vpr/src/route/check_rr_graph.cpp
Comment (litghost) [inline vpr/src/route/check_rr_graph.cpp]: These changes are not related to your PR, please remove them

**Review by litghost — CHANGES_REQUESTED** [review]

File: vpr/src/route/rr_graph_reader.cpp
Comment (litghost) [inline vpr/src/route/rr_graph_reader.cpp]: Please factor this into a function, rather than inline.

File: vpr/src/route/check_rr_graph.cpp
Comment (andrewb1999) [reply vpr/src/route/check_rr_graph.cpp]: I don't believe they are exactly related but when we switch to the check_rr_graph.cpp on master vpr crashes on checking the rr_graph we produced.

**Review by litghost — CHANGES_REQUESTED** [review]
Please do not use strlen, strdup, strcpy, etc.

Prefer std::string, std::string.substr, etc

File: vpr/src/route/check_rr_graph.cpp
Comment (litghost) [reply vpr/src/route/check_rr_graph.cpp]: See https://github.com/SymbiFlow/symbiflow-arch-defs/pull/830

Comment (andrewb1999) [reply vpr/src/route/check_rr_graph.cpp]: So is this check expected to fail now?  Or did something change such that the rest of our code no longer works? (I reverted back to the check_rr_graph.cpp on master for now)

Comment (litghost) [reply vpr/src/route/check_rr_graph.cpp]: Ya, in SymbiFlow/symbiflow-arch-defs#830 we removed the need for these capacity = 0 nodes.

Comment (andrewb1999) [reply vpr/src/route/check_rr_graph.cpp]: Thanks, updating symbiflow-arch-defs fixed this issue

File: vpr/src/route/rr_graph_reader.cpp
Comment (andrewb1999) [reply vpr/src/route/rr_graph_reader.cpp]: I believe I have made both of these changes.  Let me know if I missed anything.

**Comment by acomodi** [comment]
@andrewb1999 can you please rebase once more as I have pushed-forced on master+wip to make CI go green again?
It would also be better if you create a local branch probably and push the features in there instead of working on master+wip. This means that you should do something like:

```
cd verilog-to-routing (your local repository)
git checkout master
git branch -b new-feature-branch
(make your changes and commit)
```

To easily get back in line with SymbiFlow/master+wip do the following (I see you opened the PR from your local master+wip branch):

```
cd verilog-to-routing
git remote add upstream https://github.com/SymbiFlow/vtr-verilog-to-routing.git
git fetch upstream
git checkout master+wip
git reset --hard upstream/master+wip
git cherry-pick <commit hashes of the changes in your local branch>
git push --force origin master+wip
```

**Comment by acomodi** [comment]
@andrewb1999 The reason I am push forcing on `master+wip` frequently is that, to keep the local fork of VtR as clean as possible it is necessary to keep all the changes on top of the master branch (this way is faster to look at what are the differences with upstream VtR)

**Comment by andrewb1999** [comment]
> @andrewb1999 can you please rebase once more as I have pushed-forced on master+wip to make CI go green again?
> It would also be better if you create a local branch probably and push the features in there instead of working on master+wip. This means that you should do something like:
> 
> ```
> cd verilog-to-routing (your local repository)
> git checkout master
> git branch -b new-feature-branch
> (make your changes and commit)
> ```
> 
> To easily get back in line with SymbiFlow/master+wip do the following (I see you opened the PR from your local master+wip branch):
> 
> ```
> cd verilog-to-routing
> git remote add upstream https://github.com/SymbiFlow/vtr-verilog-to-routing.git
> git fetch upstream
> git checkout master+wip
> git reset --hard upstream/master+wip
> git cherry-pick <commit hashes of the changes in your local branch>
> git push --force origin master+wip
> ```
Alright I did this.

**Comment by acomodi** [comment]
@andrewb1999  Great, the only thing left is code formatting to let CI go green:
Just run the following:
```
make format
```

It will automatically format the codebase and you will have to commit the changes.

**Comment by andrewb1999** [comment]
> @andrewb1999 Great, the only thing left is code formatting to let CI go green:
> Just run the following:
> 
> ```
> make format
> ```
> 
> It will automatically format the codebase and you will have to commit the changes.

Done.

File: vpr/src/route/rr_graph_reader.cpp
Comment (mithro) [inline vpr/src/route/rr_graph_reader.cpp]:
```suggestion
 * seg_id is read in from the rr graph */
```

File: vpr/src/route/rr_graph.cpp
Comment (mithro) [inline vpr/src/route/rr_graph.cpp]: `al data`? Did you mean `all data`?

File: vpr/src/route/rr_graph.cpp
Comment (mithro) [inline vpr/src/route/rr_graph.cpp:2514]: Please add a description of the file format to this comment.

File: vpr/src/route/rr_graph.cpp
Comment (mithro) [inline vpr/src/route/rr_graph.cpp]: I think there might be a function like this somewhere in the code already? vpr is already produces an `rr_graph.echo` file which looks like;
```
RR node: 0 type: SOURCE location: (0,1) class: 0
1 edge(s): 1
Switch types: 0
Capacity: 1
R: 0  C: 0
Cost_index: 0

RR node: 1 type: OPIN location: (0,1) pin: 0 pin_name: IBUF.I[0]
2 edge(s): 38 39
Switch types: 1 1
Capacity: 1
R: 0  C: 0
Cost_index: 2
```

File: vpr/src/route/rr_graph.cpp
Comment (mithro) [inline vpr/src/route/rr_graph.cpp:2516]: Please add a header to the binary file which has some data like;
 - [ ] [Magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)#Magic_numbers_in_files) / [File signatures.](https://en.wikipedia.org/wiki/List_of_file_signatures) 
 - [ ] A binary file version format.
 - [ ] The details in the rr_graph.xml header (`<rr_graph tool_name="vpr" tool_version="8.0.0-dev+vpr-7.0.5-6080-gc64882fe2" tool_comment="Generated from arch file /home/tansell/github/SymbiFlow/symbiflow-arch-defs/ice40/build/ice40-top-routing-virt-hx8k/arch.xml">`)

**Review by mithro — COMMENTED** [review]
Hi @andrewb1999,

Firstly, thank you for sending this pull request! This type of functionality is something that has been on our todo list and definitely something want to eventually support in the future!

I'm pondering what the best path forward is. 

The better long term solution is to use something like flat buffers (or other existing binary) rather than a custom system. @litghost has an old "work in progress" version of this in his [`flatbuffer_rr_graph`](https://github.com/litghost/vtr-verilog-to-routing/tree/flatbuffer_rr_graph) and [`flatbuffer_rr_graph_rebase2`](https://github.com/litghost/vtr-verilog-to-routing/tree/flatbuffer_rr_graph_rebase2) branches but he has not had time to finish them off.

If you have time to take @litghost changes and make them suitable for merging upstream, that would be the best long term option. I'm happy to work with you to make that happen.

If you don't have time to do that right now, I *think* we could keep this change in the [symbiflow/vtr-verilog-to-routing fork](https://github.com/SymbiFlow/vtr-verilog-to-routing) until we can rewrite it using something like flat buffers and get it into the [upstream Verilog To Routing repository](https://github.com/verilog-to-routing/vtr-verilog-to-routing).

Thoughts?

Tim '@mithro' Ansell

File: vpr/src/route/rr_graph_reader.cpp
Comment (acomodi) [inline vpr/src/route/rr_graph_reader.cpp]: IMO here it would be better to use something like as VTR offers ways to throw exceptions and has a built in error handling, instead of using `exit(1)`:
```
VPR_THROW(VPR_ERROR_OTHER, "Binary File %s Does Not Exist\n", filename);
```

File: vpr/src/route/rr_graph_writer.cpp
Comment (acomodi) [inline vpr/src/route/rr_graph_writer.cpp:42]: It would be better probably to use C++ style for writing on files.
Something like:
```
fstream fb;
...
fb.open(bin_file_name, fstream::out | fstream::trunc);
fb << "Things to write to file..."
```

For reference, all the other functions that do write information about an RR_graph make use of the same methodology.

File: vpr/src/route/rr_graph_writer.cpp
Comment (acomodi) [inline vpr/src/route/rr_graph_writer.cpp]: I suppose this is a log print. If I am correct IMO it would be best to use the VTR built-in logging mechanism:
```
VTR_LOG("RR_graph binary file: |%s|\n", bin_file_name.c_str());
```

File: vpr/src/route/rr_graph_writer.cpp
Comment (acomodi) [inline vpr/src/route/rr_graph_writer.cpp]: Same as above:
```
VTR_LOG("RR_graph binary mode\n");
```

**Review by acomodi — CHANGES_REQUESTED** [review]

**Comment by andrewb1999** [comment]
> Hi @andrewb1999,
> 
> Firstly, thank you for sending this pull request! This type of functionality is something that has been on our todo list and definitely something want to eventually support in the future!
> 
> I'm pondering what the best path forward is.
> 
> The better long term solution is to use something like flat buffers (or other existing binary) rather than a custom system. @litghost has an old "work in progress" version of this in his [`flatbuffer_rr_graph`](https://github.com/litghost/vtr-verilog-to-routing/tree/flatbuffer_rr_graph) and [`flatbuffer_rr_graph_rebase2`](https://github.com/litghost/vtr-verilog-to-routing/tree/flatbuffer_rr_graph_rebase2) branches but he has not had time to finish them off.
> 
> If you have time to take @litghost changes and make them suitable for merging upstream, that would be the best long term option. I'm happy to work with you to make that happen.
> 
> If you don't have time to do that right now, I _think_ we could keep this change in the [symbiflow/vtr-verilog-to-routing fork](https://github.com/SymbiFlow/vtr-verilog-to-routing) until we can rewrite it using something like flat buffers and get it into the [upstream Verilog To Routing repository](https://github.com/verilog-to-routing/vtr-verilog-to-routing).
> 
> Thoughts?
> 
> Tim '@mithro' Ansell

Okay I have talked to Andre about this and we agree that making a more permanent solution with flat buffers is not our priority right now (we're focusing on DSPs and design verification).  I am willing to spend some time to cleanup this code if it would be a useful placeholder in the SymbiFlow vtr fork until someone can get around to a solution with flat buffers.  It would be useful for us to have this code integrated into the symbiflow fork to make it easier for us to stay up to date with other VTR development.  Let me know your thoughts.

Andrew

**Comment by acomodi** [comment]
@andrewb1999 Hi,
I have push-forced once again to include other patches in VTR, could you do the same steps you did [here](https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/120#issuecomment-504620347)? .

I think that this PR can be merged once the review comments have been addressed. As soon as those minor issues have been solved, the PR is rebased and CI goes green we could merge in `master+wip`.

It would be great if you could also squash all the commits in a single one with the patch. I will proceed then by creating a local branch in SymbiFlow/VtR fork with this patch and merge it into the octopus merge in `master+wip` with all the other pending branches until binary read/write mode is integrated in upstream VtR.

**Comment by mithro** [comment]
> Okay I have talked to Andre about this and we agree that making a more permanent solution with flat buffers is not our priority right now (we're focusing on DSPs and design verification). I am willing to spend some time to cleanup this code if it would be a useful placeholder in the SymbiFlow vtr fork until someone can get around to a solution with flat buffers. It would be useful for us to have this code integrated into the symbiflow fork to make it easier for us to stay up to date with other VTR development. Let me know your thoughts.

If it keeps you on the current head of SymbiFlow, then after a little more clean up, I'm happy to have this branch in the SymbiFlow VtR version until we land a replacement which uses a better format (this branch will then be deleted).

File: vpr/src/route/rr_graph_writer.cpp
Comment (andrewb1999) [reply vpr/src/route/rr_graph_writer.cpp:42]: As far as I can tell this won't work for binary.  It will write out as ascii text which is not what we want.  There's a possibility of changing over to fstream write() but I do not this this is much different from the current fwrite() and doesn't seem to be worth it for a temporary solution.

Comment (acomodi) [reply vpr/src/route/rr_graph_writer.cpp:42]: Yeah, I believe you are right, using fstream would overcomplicate the solution to be able to write in bianry. As you say I think it could be kept as is for now.

**Comment by andrewb1999** [comment]
Made recommended changes and squashed commits.

**Comment by acomodi** [comment]
@andrewb1999 Thanks for the changes. I think that I can proceed by creating a new branch with your changes and add it to the octopus merge alongside with the other branches.

@litghost, @mithro If this is ok to you as well I will include this patch in the next octopus merge.

**Review by acomodi — APPROVED** [review]

File: vpr/src/route/rr_graph_reader.cpp
Comment (acomodi) [inline vpr/src/route/rr_graph_reader.cpp]: @andrewb1999  Just noticed this. Is it necessary to comment this check? If not it would be better to de-comment it.

Comment (andrewb1999) [reply vpr/src/route/rr_graph_reader.cpp]: That was a mistake (although we are planning on adding an option to disable it eventually).  Should be fixed now.

File: vpr/src/route/rr_graph_reader.cpp
Comment (acomodi) [inline vpr/src/route/rr_graph_reader.cpp]: One more thing from my side. I have tried to compile VPR with this patch and numerous warnings, related to the unused result of `fread`, showed up.
There are two ways to solve this:
- Disable warnings related to unused results (which I am not fund of);
- Add a wrapper helper function (to [`vpr_utils.cpp`](https://github.com/SymbiFlow/vtr-verilog-to-routing/blob/master%2Bwip/vpr/src/util/vpr_utils.cpp) for example) that checks the result of `fread` and, in case, throws an error. It would be something like this:

```
 void fread_secure(void *var, size_t size, unsigned int count, FILE *fp) {                                                                                                                                                                                                                
     auto result = fread(var, size, count, fp);                                                                                                                                                                                                                                           
     if (result != count) {                                                                                                                                                                                                                                                               
         VPR_THROW(VPR_ERROR_OTHER, "ERROR reading file\n");                                                                                                                                                                                                                                
     }                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                          
     return;                                                                                                                                                                                                                                                                              
 }  
``` 
Then change all the `fread` into `fread_secure` and this solves the warning problem and makes the solution more robust IMO.

Comment (andrewb1999) [reply vpr/src/route/rr_graph_reader.cpp]: Done

**Comment by acomodi** [comment]
@andrewb1999 I have created a new branch with this patch: https://github.com/SymbiFlow/vtr-verilog-to-routing/tree/rr-graph-binary-read-write and merged it within `master+wip`.

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/SymbiFlow/vtr-verilog-to-routing">SymbiFlow/vtr-verilog-to-routing</a></strong></div><div>SymbiFlow WIP changes for Verilog to Routing -- Open Source CAD Flow for FPGA Research - SymbiFlow/vtr-verilog-to-routing</div></blockquote>

**Comment by litghost** [comment]
This PR is superseded by https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1060 and https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/340

**Pull request closed without merging**
